sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertSuite.scala

```diff
@@ -847,4 +847,26 @@ class InsertSuite extends QueryTest with TestHiveSingleton with BeforeAndAfter
       }
     }
   }
+
+  test("SPARK-32508 " +
+    "Disallow empty part col values in partition spec before static partition writing") {
+    withTable("t1") {
+      spark.sql(
+        """
+          |CREATE TABLE t1 (c1 int)
```

Comment:
why is this a problem only for hive tables?

Reply:
`InsertIntoHadoopFsRelationCommand`
When `manageFilesourcePartitions` is turned on,`catalog.listPartitions` is called, here is a check to see if the partition value is empty.

In the case that `manageFilesourcePartitions` is not turned on, the partition value is currently not checked, which means that the SQL execution will not fail. If I now move the check logic to the PreprocessTableInsertion rule, this will cause the execution to fail.

Perhaps this check can only be performed when `tracksPartitionsInCatalog` is equal to true and the static partition is written.

Reply:
hive calls `getPartition` when `loadPartition`, here it will check whether the partition value is empty.

```java
public Partition getPartition(...){
          || (val != null && val.length() == 0)) {
        throw new HiveException("get partition: Value for key "
            + field.getName() + " is null or empty");
}
```

Reply:
> In the case that manageFilesourcePartitions is not turned on, the partition value is currently not checked, which means that the SQL execution will not fail.

what's the behavior then?

Reply:
`InsertIntoHadoopFsRelationCommand`

When writing static partition or  dynamic partition, the `DynamicPartitionDataWriter` will be used, and the partition value of empty will generate the default value(__HIVE_DEFAULT_PARTITION__) through `getPartitionPathString`.

When `manageFilesourcePartitions` is not turned on, the partition information is maintained through the filesystem, so it is not checked whether the partition value is empty.
`insert ovwriter table a partition(d='') select 1` sql will run successfully.

Reply:
what's the behavior of hive? Does hive treat `d=''` as `d= HIVE_DEFAULT_PARTITION` and can run it?

Reply:
hive> INSERT OVERWRITE TABLE  t1 PARTITION(d='')  select 1;
```
Caused by: org.apache.hadoop.hive.ql.parse.SemanticException: Line 1:61 Partition not found ''''
        at org.apache.hadoop.hive.ql.parse.BaseSemanticAnalyzer$tableSpec.<init>(BaseSemanticAnalyzer.java:856)
        at org.apache.hadoop.hive.ql.parse.BaseSemanticAnalyzer$tableSpec.<init>(BaseSemanticAnalyzer.java:727)
        at org.apache.hadoop.hive.ql.parse.SemanticAnalyzer.getMetaData(SemanticAnalyzer.java:1652)
        ... 23 more
Caused by: org.apache.hadoop.hive.ql.metadata.HiveException: get partition: Value for key d is null or empty
        at org.apache.hadoop.hive.ql.metadata.Hive.getPartition(Hive.java:1900)
```

hive> INSERT OVERWRITE TABLE  t1 PARTITION(d)  select 1,'' as d;
result:
```
Loading data to table x.t1 partition (d=null)
	 Time taken for load dynamic partitions : 302
	Loading partition {d=__HIVE_DEFAULT_PARTITION__}
	 Time taken for adding to write entity : 1
Partition x.t1{d=__HIVE_DEFAULT_PARTITION__} stats: [numFiles=1, numRows=1, totalSize=201, rawDataSize=85]

```